File: arrays_hashing/valid_anagram.py
class Solution:
    """
    Problem: Given two strings s and t, return true if t is an anagram of s, and false otherwise.
    An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase,
    typically using all the original letters exactly once.
    """
    def isAnagram(self, s: str, t: str) -> bool:
        if s is None or t is None:
            return False
        
        if len(s) != len(t):
            return False
            
        # Create Frequency Map
        frequency_map = {}
        
        for c in s:
            frequency_map[c] = frequency_map.get(c, 0) + 1
            
        # Loop through string t and lookup in frequency map
        for c in t:
            if c not in frequency_map:
                return False
            else:
                count = frequency_map[c]
                if count == 0:
                    return False
                frequency_map[c] = count - 1
                
        return True

if __name__ == "__main__":
    solution = Solution()
    
    # Test Case 1
    s1, t1 = "anagram", "nagaram"
    print(f"Test Case 1: {solution.isAnagram(s1, t1)}") # Expected: True

    # Test Case 2
    s2, t2 = "rat", "car"
    print(f"Test Case 2: {solution.isAnagram(s2, t2)}") # Expected: False

File: arrays_hashing/ValidAnagram.java
package arrays_hashing;

import java.util.*;

public class ValidAnagram {
    
    // Problem: Given two strings s and t, return true if t is an anagram of s, and false otherwise.
    // An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase,
    // typically using all the original letters exactly once.

    public boolean isAnagram(String s, String t) {
            if(s == null || t == null)
                return false;
            if(s.length() != t.length()){
                return false;
            }
            // Create Frequency Map
            Map<Character, Integer> frequencyMap = new HashMap<>();
            for(char c: s.toCharArray()){
                frequencyMap.put(c, frequencyMap.getOrDefault(c, 0) + 1);
            }
    
            // Loop through string t and lookup in frequency map
            for(char c: t.toCharArray()) {
                if(!frequencyMap.containsKey(c)) {
                    return false; 
                } else {
                    int count = frequencyMap.get(c);
                    if(count == 0) {
                        return false;
                    }
                    frequencyMap.put(c, count-1);    
                }
            }
    
            // Loop through map and verify the count values are all zeroes
            for(int count: frequencyMap.values()) {
                    if(count > 0) {
                        return false;
                    }
            } 
            return true;
            
        }
    

    public static void main(String[] args) {
        ValidAnagram solution = new ValidAnagram();
        
        // Test Case 1
        String s1 = "anagram", t1 = "nagaram";
        System.out.println("Test Case 1: " + solution.isAnagram(s1, t1)); // Expected: true

        // Test Case 2
        String s2 = "rat", t2 = "car";
        System.out.println("Test Case 2: " + solution.isAnagram(s2, t2)); // Expected: false
    }
}

